src/Spectre.libGaussianMixtureModelling/MaximizationRunnerOpt.h

```diff
@@ -0,0 +1,125 @@
+*/
+#pragma once
+#include <vector>
+#include "DataTypes.h"
+#include "GaussianMixtureModel.h"
+#include "Matrix.h"
+
+namespace spectre::unsupervised::gmm
+{
+    /// <summary>
+    /// Class serves the purpose of maximization step of Expectation
+    /// Maximization algorithm.
+    /// </summary>
+    class MaximizationRunnerOpt
+    {
+    public:
+        /// <summary>
+        /// Constructor initializing the class with data required during
+        /// maximization step.
+        /// </summary>
+        /// <param name="spectrum">Spectrum to base calculations on.</param>
+        /// <param name="affilationMatrix">Matrix symbolising the probability
+        /// of affilation of each sample to a certain gaussian component.
+        /// </param>
+        /// <param name="components">Gaussian components to be updated</param>
+        MaximizationRunnerOpt(SpectrumView spectrum,
+            Matrix<DataType> &affilationMatrix,
+            std::vector<GaussianComponent> &components)
+            : m_Spectrum(spectrum), m_AffilationMatrix(affilationMatrix)
+            , m_Components(components)
+            , m_SummedIntensities(std::accumulate(spectrum.intensities.begin(),
+                spectrum.intensities.end(), 0.0))
+        {
+        }
+
+        /// <summary>
+        /// Contratry to what name indicates, it updates weights, means and
+        /// standard deviations in a single run.
+        /// </summary>
+        void UpdateWeights()
+        {
+            const unsigned numberOfComponents = (unsigned)m_Components.size();
+            const unsigned dataSize = (unsigned)m_Spectrum.mzs.size();
+            DataView mzs = m_Spectrum.mzs;
+            for (unsigned k = 0; k < numberOfComponents; k++)
+            {
+                DataType probabilitySum = 0.0;
+                DataType meanNumerator = 0.0;
+                for (unsigned i = 0; i < (unsigned)m_Spectrum.mzs.size(); i++)
+                {
+                    probabilitySum += m_AffilationMatrix[k][i];
+                    meanNumerator += m_AffilationMatrix[k][i] * mzs[i];
+                }
+                m_Components[k].weight = probabilitySum / m_SummedIntensities;
+                m_Components[k].mean = meanNumerator / probabilitySum;
+                DataType stdNumerator = 0.0;
+                for (unsigned i = 0; i < (unsigned)m_Spectrum.mzs.size(); i++)
+                {
+                    DataType diff = mzs[i] - m_Components[k].mean;
+                    stdNumerator += m_AffilationMatrix[k][i] * diff * diff;
+                }
+                m_Components[k].deviation = sqrt(stdNumerator/probabilitySum);
+            }
+        }
+
+        /// <summary>
+        /// Empty function made to conform with interface.
+        /// </summary>
+        void UpdateMeans()
+        {
+        }
+
+        /// <summary>
+        /// Empty function made to conform with interface.
+        /// </summary>
+        void UpdateStdDeviations()
```

Comment:
Are these functions meant to be filled in the future? If they are only to conform with interface why are they even called anywhere?

Reply:
Due to existing API this cannot be changed without refining everything else. 
To be refactored in the future.

Reply:
Such approach is bad. If API is unsuitable, make it fit. It's better to do it now, than to carry it longer, to let problem grow.